File: src/backend/app/schemas/resilience.py
"""Pydantic schemas for the Resilience Wallet and RRI API."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


# ── Dimension schemas ─────────────────────────────────────────────────────────

class DimensionStateSchema(BaseModel):
    dimension: str
    maximum_balance: float
    remaining_balance: float
    consumed_balance: float
    utilization_percent: float
    dimension_score: float
    status: str
    unit: str


class WalletStateSchema(BaseModel):
    shipment_id: int
    shipment_code: str
    time: DimensionStateSchema
    cost: DimensionStateSchema
    temperature: DimensionStateSchema
    capacity: DimensionStateSchema
    is_resilience_bankrupt: bool
    computed_at: datetime


# ── RRI schemas ───────────────────────────────────────────────────────────────

class ContributorSchema(BaseModel):
    factor: str
    impact: float
    reason: str
    severity: str


class RRIExplanationSchema(BaseModel):
    rri: float
    status: str
    dimensions: dict[str, float]
    weights: dict[str, float]
    weighted_scores: dict[str, float]
    contributors: list[ContributorSchema]
    criticality_adjustment: float
    weighted_score_before_adjustment: float
    bankruptcy_flags: dict[str, bool]
    is_resilience_bankrupt: bool
    shipment_priority: str
    computed_at: str


# ── Combined response ─────────────────────────────────────────────────────────

class ResilienceResponse(BaseModel):
    wallet: WalletStateSchema
    rri: RRIExplanationSchema


# ── Transactions ──────────────────────────────────────────────────────────────

class TransactionSchema(BaseModel):
    id: int
    wallet_id: int
    dimension: str
    amount: float
    reason: str
    reference_id: str | None
    timestamp: datetime

    model_config = {"from_attributes": True}


# ── Apply-event request ───────────────────────────────────────────────────────

class ApplyEventRequest(BaseModel):
    dimension: Literal["time", "cost", "temperature", "capacity"]
    amount: float = Field(gt=0, description="Amount to consume (must be positive)")
    reason: str = Field(min_length=3, max_length=500)
    reference_id: str | None = Field(default=None, max_length=100)

    @field_validator("amount")
    @classmethod
    def amount_must_be_finite(cls, v: float) -> float:
        import math
        if not math.isfinite(v):
            raise ValueError("amount must be a finite number")
        return v


# ── Dashboard aggregate ───────────────────────────────────────────────────────

class RRISummarySchema(BaseModel):
    total_shipments_assessed: int
    average_rri: float
    healthy_count: int
    stressed_count: int
    vulnerable_count: int
    critical_count: int
    bankrupt_count: int
    resilience_bankruptcies: int
    computed_at: datetime

File: src/frontend/src/api/resilience.ts
/**
 * Resilience API client — typed wrappers for Phase 3 endpoints.
 */

const BASE_URL: string = import.meta.env.VITE_API_BASE_URL ?? ''

export interface DimensionState {
  dimension: string
  maximum_balance: number
  remaining_balance: number
  consumed_balance: number
  utilization_percent: number
  dimension_score: number
  status: 'HEALTHY' | 'STRESSED' | 'VULNERABLE' | 'CRITICAL' | 'BANKRUPT'
  unit: string
}

export interface WalletState {
  shipment_id: number
  shipment_code: string
  time: DimensionState
  cost: DimensionState
  temperature: DimensionState
  capacity: DimensionState
  is_resilience_bankrupt: boolean
  computed_at: string
}

export interface Contributor {
  factor: string
  impact: number
  reason: string
  severity: string
}

export interface RRIExplanation {
  rri: number
  status: 'HEALTHY' | 'STRESSED' | 'VULNERABLE' | 'CRITICAL' | 'BANKRUPT'
  dimensions: Record<string, number>
  weights: Record<string, number>
  weighted_scores: Record<string, number>
  contributors: Contributor[]
  criticality_adjustment: number
  weighted_score_before_adjustment: number
  bankruptcy_flags: Record<string, boolean>
  is_resilience_bankrupt: boolean
  shipment_priority: string
  computed_at: string
}

export interface ResilienceResponse {
  wallet: WalletState
  rri: RRIExplanation
}

export interface ResilienceTransaction {
  id: number
  wallet_id: number
  dimension: string
  amount: number
  reason: string
  reference_id: string | null
  timestamp: string
}

export interface RRISummary {
  total_shipments_assessed: number
  average_rri: number
  healthy_count: number
  stressed_count: number
  vulnerable_count: number
  critical_count: number
  bankrupt_count: number
  resilience_bankruptcies: number
  computed_at: string
}

export interface ApplyEventRequest {
  dimension: 'time' | 'cost' | 'temperature' | 'capacity'
  amount: number
  reason: string
  reference_id?: string
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${BASE_URL}${path}`, {
    headers: { 'Content-Type': 'application/json', ...init?.headers },
    ...init,
  })
  if (!res.ok) {
    const detail = await res.text()
    throw new Error(`API ${res.status}: ${detail}`)
  }
  return res.json() as Promise<T>
}

export const resilienceApi = {
  getResilience: (shipmentId: number) =>
    request<ResilienceResponse>(`/api/shipments/${shipmentId}/resilience`),

  recalculate: (shipmentId: number) =>
    request<ResilienceResponse>(`/api/shipments/${shipmentId}/resilience/recalculate`, {
      method: 'POST',
    }),

  applyEvent: (shipmentId: number, event: ApplyEventRequest) =>
    request<ResilienceResponse>(`/api/shipments/${shipmentId}/resilience/apply-event`, {
      method: 'POST',
      body: JSON.stringify(event),
    }),

  getTransactions: (shipmentId: number) =>
    request<ResilienceTransaction[]>(`/api/shipments/${shipmentId}/resilience/transactions`),

  getSummary: () =>
    request<RRISummary>('/api/shipments/resilience/summary'),
}
